src/vos/vos_query.c

```diff
@@ -255,113 +222,282 @@ _query_recx(struct open_query *query, daos_recx_t *recx, struct evt_extent *filt
+			rc = prc;
+			break;
+		}
+
+		if (nrc == 0 && nrefresh) {
+			rc = evt_iter_fetch(nih, &inob, &nentry, NULL);
+			if (rc != 0)
+				break;
+
+			nrefresh = false;
+		}
+		if (prc == 0 && prefresh) {
+			rc = evt_iter_fetch(pih, &inob, &pentry, NULL);
+			if (rc != 0)
+				break;
+
+			/* Fake the parity bounds to match the equivalent data range */
+			ext = pentry.en_sel_ext;
+			ext.ex_lo = ext.ex_lo ^ DAOS_EC_PARITY_BIT;
+			ext.ex_hi = ext.ex_hi ^ DAOS_EC_PARITY_BIT;
+			ext.ex_lo = ext.ex_lo * cells_per_stripe;
+			ext.ex_hi = (ext.ex_hi + 1) * cells_per_stripe - 1;
+			pentry.en_sel_ext = ext;
+			prefresh = false;
+		}
+
+		D_ASSERT(prc == 0 || prc == -DER_NONEXIST);
+		D_ASSERT(nrc == 0 || nrc == -DER_NONEXIST);
+
+		if (prc == -DER_NONEXIST) {
+			if (bio_addr_is_hole(&nentry.en_addr)) {
+				nrefresh = true;
+				goto next;
+			}
+
+			/** There is no parity and the data entry isn't a hole, so return it */
+			ent2recx(recx, &nentry);
+			break;
+		} else if (nrc == -DER_NONEXIST) {
+			/** Use the adjusted parity extent */
+			ent2recx(recx, &pentry);
+			break;
+		}
+
+		/** Ok, they both exist, so we need to determine which one meets the criteria or
+		 *  continue the search for until we find one or run out of options.
+		 */
+		if (query->qt_flags & VOS_GET_MAX) {
+			if (find_answer(&pentry, &nentry, recx, &nrefresh, &prefresh, hi_gtr,
+					is_after, overlap_lo))
+				break;
+		} else {
+			if (find_answer(&pentry, &nentry, recx, &nrefresh, &prefresh, lo_lt,
+					is_before, overlap_hi))
+				break;
+		}
+next:
+		if (nrefresh)
+			nrc = evt_iter_next(nih);
+		if (prefresh)
+			prc = evt_iter_next(pih);
 	}
 
-	/* possible punched recxs within this parity recx */
-	filter_ex.ex_lo = recxs[1].rx_idx;
-	D_ASSERT((filter_ex.ex_lo & DAOS_EC_PARITY_BIT) != 0);
-	filter_ex.ex_lo &= ~DAOS_EC_PARITY_BIT;
-	D_ASSERT(filter_ex.ex_lo % query->qt_cell_size == 0);
-	filter_ex.ex_hi = (((filter_ex.ex_lo + recxs[1].rx_nr) / query->qt_cell_size) *
-			   query->qt_stripe_size);
-	filter_ex.ex_lo = ((filter_ex.ex_lo / query->qt_cell_size) *
-			   query->qt_stripe_size);
-
-	rc = query_ec_normal_recx(query, &recxs[2], &filter_ex, false);
+	D_DEBUG(DB_TRACE, "query recx "DF_U64"/"DF_U64" : "DF_RC"\n", recx->rx_idx,
+		recx->rx_nr, DP_RC(rc));
+	if (rc == 0)
+		exist = true;
 	if (rc == -DER_NONEXIST)
 		rc = 0;
-
-	D_DEBUG(DB_TRACE, "get punched recxs "DF_U64"/"DF_U64"\n", recxs[2].rx_idx,
-		recxs[2].rx_nr);
+	close_rc = evt_iter_finish(pih);
+	if (rc == 0)
+		rc = close_rc;
+close_nih:
+	close_rc = evt_iter_finish(nih);
+	if (rc == 0)
+		rc = close_rc;
+out:
+	close_rc = evt_close(toh);
+	if (rc == 0 && !exist)
+		rc = -DER_NONEXIST;
+	if (rc == 0)
+		rc = close_rc;
 
 	return rc;
 }
 
+static int
+query_recx(struct open_query *query, daos_recx_t *recxs)
+{
+	if (!(query->qt_flags & VOS_GET_RECX_EC))
+		return query_normal_recx(query, &recxs[0]);
+
+	memset(&recxs[1], 0, sizeof(recxs[1]) * 2);
+	return query_ec_recx(query, &recxs[0]);
```

Comment:
Note that we now only need and return a single recx here but I didn't really want to mess with the protocol.